### ISSUE TITLE
Different rendering between racket-mode and DrRacket

### ISSUE THREAD

**Issue by sturgman** [greghendershott/racket-mode#397]
Hello! I am following the getting started guide here: https://docs.racket-lang.org/quick/index.html and noticed that Racket-mode and DrRacket give different output for the command `(rainbow (square 5))` . In emacs I get this:

![image](https://user-images.githubusercontent.com/4121024/63052087-836b2400-beac-11e9-8c3f-3760dd23f2f2.png)

Note the newlines after each square. I couldn't find this in the issue list so I wasn't sure if it was known behavior. For convenience, here are the definitions from the tutorial:

```
#lang slideshow
(define (square n)
  ; Renders a black square
  (filled-rectangle n n)
  )

(define (four p)
  (define two-p (hc-append p p))
  (vc-append two-p two-p))

(define (checker p1 p2)
  (let ([p12 (hc-append p1 p2)]
        [p21 (hc-append p2 p1)])
    (vc-append p12 p21)))

(define (checkerboard p)
  (let* ([rp (colorize p "red")]
         [bp (colorize p "black")]
         [c (checker rp bp)]
         [c4 (four c)])
    (four c4)))

(define (series mk)
  (hc-append 4 (mk 5) (mk 10) (mk 20)))

(define (rgb-series mk)
  (vc-append
   (series (lambda (sz) (colorize (mk sz) "red")))
   (series (lambda (sz) (colorize (mk sz) "green")))
   (series (lambda (sz) (colorize (mk sz) "blue")))))

(define (rgb-maker mk)
  (lambda (sz)
    (vc-append (colorize (mk sz) "red")
               (colorize (mk sz) "green")
               (colorize (mk sz) "blue"))))

(define (rainbow p)
  (map (lambda (color)
         (colorize p color))
       (list "red" "orange" "yellow" "green" "blue" "purple")))
```

<details>
```
((alist-get 'racket-mode package-alist)
 #s(package-desc racket-mode
		 (20190803 1820)
		 "Major mode for Racket language."
		 ((emacs
		   (24 3))
		  (faceup
		   (0 0 2)))
		 nil nil "c:/Users/sturgmancohen/.emacs.d/elpa/racket-mode-20190803.1820"
		 ((:url . "https://www.racket-mode.com/")
		  (:maintainer "Greg Hendershott")
		  (:authors
		   ("Greg Hendershott")))
		 nil))
((emacs-version "26.1")
 (emacs-uptime "36 minutes, 56 seconds")
 (system-type windows-nt)
 (major-mode racket-repl-mode)
 (racket--el-source-dir "c:/Users/sturgmancohen/.emacs.d/elpa/racket-mode-20190803.1820/")
 (racket--rkt-source-dir "c:/Users/sturgmancohen/.emacs.d/elpa/racket-mode-20190803.1820/racket/")
 (racket-program "Racket.exe")
 (racket-command-port 55555)
 (racket-command-startup nil)
 (racket-command-timeout 10)
 (racket-memory-limit 2048)
 (racket-error-context medium)
 (racket-retry-as-skeleton t)
 (racket-error-context medium)
 (racket-history-filter-regexp "\\`\\s *\\S ?\\S ?\\s *\\'")
 (racket-images-inline t)
 (racket-images-keep-last 100)
 (racket-use-repl-submit-predicate nil)
 (racket-images-system-viewer "display")
 (racket-pretty-print t)
 (racket-indent-curly-as-sequence t)
 (racket-indent-sequence-depth 0)
 (racket-pretty-lambda nil)
 (racket-smart-open-bracket-enable nil)
 (racket-module-forms "\\s(\\(?:module[*+]?\\|library\\)")
 (racket-logger-config
  ((cm-accomplice . warning)
   (GC . info)
   (module-prefetch . warning)
   (optimizer . info)
   (racket/contract . error)
   (sequence-specialization . info)
   (* . fatal))))
(enabled-minor-modes
 (TeX-PDF-mode)
 (TeX-source-correlate-mode)
 (async-bytecomp-package-mode)
 (auto-composition-mode)
 (auto-compression-mode)
 (auto-encryption-mode)
 (auto-fill-mode)
 (auto-insert-mode)
 (auto-save-mode)
 (blink-cursor-mode)
 (csv-field-index-mode)
 (desktop-save-mode)
 (electric-indent-mode)
 (file-name-shadow-mode)
 (font-lock-mode)
 (global-eldoc-mode)
 (global-flycheck-mode)
 (global-font-lock-mode)
 (helm-mode)
 (line-number-mode)
 (menu-bar-mode)
 (mouse-wheel-mode)
 (override-global-mode)
 (persistent-scratch-autosave-mode)
 (recentf-mode)
 (shell-dirtrack-mode)
 (tooltip-mode)
 (transient-mark-mode)
 (yas-global-mode)
 (yas-minor-mode))
(disabled-minor-modes
 (LaTeX-math-mode)
 (TeX-Omega-mode)
 (TeX-interactive-mode)
 (abbrev-mode)
 (auto-fill-function)
 (auto-save-visited-mode)
 (buffer-read-only)
 (cl-old-struct-compat-mode)
 (column-number-mode)
 (company-mode)
 (company-search-mode)
 (compilation-in-progress)
 (compilation-minor-mode)
 (compilation-shell-minor-mode)
 (completion-in-region-mode)
 (conda-env-autoactivate-mode)
 (defining-kbd-macro)
 (dired-hide-details-mode)
 (display-time-mode)
 (doc-view-minor-mode)
 (edebug-mode)
 (eldoc-mode)
 (electric-layout-mode)
 (electric-pair-mode)
 (electric-quote-mode)
 (flycheck-mode)
 (flymake-mode)
 (flyspell-mode)
 (global-company-mode)
 (global-prettify-symbols-mode)
 (global-visual-line-mode)
 (gnus-dead-summary-mode)
 (gnus-undo-mode)
 (helm--minor-mode)
 (helm--remap-mouse-mode)
 (helm-autoresize-mode)
 (helm-ff--delete-async-modeline-mode)
 (helm-migemo-mode)
 (helm-popup-tip-mode)
 (horizontal-scroll-bar-mode)
 (hs-minor-mode)
 (ido-everywhere)
 (image-minor-mode)
 (isearch-mode)
 (ispell-minor-mode)
 (jit-lock-debug-mode)
 (mail-abbrevs-mode)
 (mml-mode)
 (next-error-follow-minor-mode)
 (org-bullets-mode)
 (org-cdlatex-mode)
 (org-indent-mode)
 (org-src-mode)
 (orgstruct-mode)
 (outline-minor-mode)
 (outorg-edit-minor-mode)
 (outshine-mode)
 (overwrite-mode)
 (paragraph-indent-minor-mode)
 (prettify-symbols-mode)
 (racket-check-syntax-mode)
 (sensitive-mode)
 (server-mode)
 (sh-electric-here-document-mode)
 (size-indication-mode)
 (temp-buffer-resize-mode)
 (tool-bar-mode)
 (unify-8859-on-decoding-mode)
 (unify-8859-on-encoding-mode)
 (url-handler-mode)
 (use-hard-newlines)
 (view-mode)
 (visible-mode)
 (visual-line-mode)
 (window-divider-mode)
 (winner-mode))
```
</details>

**Comment by greghendershott** [comment]
A quick lazy guess (I don't have time right now to confirm):

The newlines are probably a result of [`racket-pretty-print`](https://www.racket-mode.com/#racket_002dpretty_002dprint) being set to `t`. You could change it to `nil`.

---

(Maybe that will answer your question and you'll close this?  Even so, with pretty-print enabled, I'm surprised by that indentation. That might be a bug and/or worth looking at as/when I have more time.)

**Comment by sturgman** [comment]
Yes! I spent some time looking at it but couldn't fix the issue in a good way. With default settings I get what I show in the original post. If I set `racket-images-inline` to `nil` I get:
```racket
'([image] ; use M-x racket-view-last-image to view
             [image] ; use M-x racket-view-last-image to view
             [image] ; use M-x racket-view-last-image to view
             [image] ; use M-x racket-view-last-image to view
             [image] ; use M-x racket-view-last-image to view
             [image] ; use M-x racket-view-last-image to view)
```
which still includes the indentation and newlines. So I ruled out the rendering of actual pictures as the issue. This `[image]` lines are written by `racket-repl--replace-images` although I don't see anything there that would do this. Then again, I really don't know what I am doing 😀 .
If I disable the pretty printer I get:
```racket
'(#<pict> #<pict> #<pict> #<pict> #<pict> #<pict>)
```
So your guess might be correct. Without the pretty printer a list shows up in the same line. Hope some of this helps when you come back to it. Thanks for the great mode!

**Comment by sturgman** [comment]
A quick update on this that may be related. I started working trough The Little Schemer and found this hideous list:
```racket
'((how much (wood)) could (( a (wood) chuck)) (((chuck))) (if (a) ((wood chuck))) could chuck wood)
```
This is how the emacs repl prints it:
![image](https://user-images.githubusercontent.com/4121024/63543896-d028ad80-c4f1-11e9-9fa2-bd0370621657.png)

This is how DrRacket renders it:
![image](https://user-images.githubusercontent.com/4121024/63543990-fe0df200-c4f1-11e9-9003-3a3e063dd9ef.png)

The DrRacket version is equivalent to the one I get in the racket terminal as well. Anyways, the difference made me think of this bug and that the problem may not be image specific.

Interestingly, this ` '((a) ((b) (c)) (a) (a) (b)  (((fdafsdddddddda)) fdkshfskh))` renders:
![image](https://user-images.githubusercontent.com/4121024/63544162-7f658480-c4f2-11e9-9bd4-f64f4826ccb6.png)

While this (with smaller internal items) `'((a) ((b) (c)) (a) (a) (b)  (((fdafsa)) fdkshfskh))` renders like so:
![image](https://user-images.githubusercontent.com/4121024/63544221-a2903400-c4f2-11e9-91e3-19a16ccbc1d5.png)

Here is in hopes that I am not driving you crazy with these. Just wanted to keep a record of the issue.

**Comment by greghendershott** [comment]
I just wanted to say, thanks for reporting these. I haven't had a chance to look more, yet.

**Comment by sturgman** [comment]
No problem! Thank you for the wonderful mode!

**Comment by greghendershott** [comment]
I notice that, if the _first_ line were indented to start in the column where the prompt ends, then the remaining lines would be indented correctly.

Racket ports can have "port positions" -- they can keep track of the current line and column. I suspect that the printer thinks it is printing at an offset column.

It is possible that commit 953eebd to fix issue #381 introduced this a few months ago -- but I have to run right now so I will investigate later.
